### ISSUE TITLE
[Question] Recovering from INVALID_PRODUCER_EPOCH

### ISSUE THREAD

**Issue by dcrodman** [twmb/franz-go#23]
Hey @twmb, over the course of testing a transactional producer I've written I've encountered this error from `EndTransaction` after a rebalance:
```
INVALID_PRODUCER_EPOCH: Producer attempted an operation with an old epoch.
```

For a bit of context, this is from a consumer (currently Sarama, will be franz-go soon) that does the following:
1. `client.BeginTransaction`
2. Consumes a record from a topic
3. Writes a bunch of records to a different topic (`client.Send`)
4. If all goes well, `client.EndTransaction(ctx, kgo.TryCommit)`
5. If something goes wrong in step 3 or 4, there's a `client.AbortBufferedRecords` followed by a `client.EndTransaction(ctx, kgo.TryAbort)`

We're running multiple instances of these consumers, each reading from one or more partitions on the input topic and writing (randomly) to partitions on the output topic. These tasks can be relatively long running and rebalances may occur during any of those steps due to deploys, scaling, kafka operations, etc.

What happens after a rebalance is they start throwing those `kerr.InvalidProducerEpoch` errors. A little digging suggests this is due to franz-go's handling of [KIP-588](https://cwiki.apache.org/confluence/display/KAFKA/KIP-588%3A+Allow+producers+to+recover+gracefully+from+transaction+timeouts) and that seems to make sense, but what isn't clear to me is how to recover from this situation. Right now, the producers just spin forever on that error - able to begin a transaction but not write records or end the transaction (though aborting still works).

I remembered this doc from `config.TransactionalID`, but I'm not sure it applies to me since I'm not using a franz-go consumer.

>// Note that, unless using Kafka 2.5.0, a consumer group rebalance may be
// problematic. Production should finish and be committed before the client
// rejoins the group. It may be safer to use an eager group balancer and just
// abort the transaction. Alternatively, any time a partition is revoked, you
// could abort the transaction and reset offsets being consumed.

So with that in mind and with the bold assumption that I'm doing this correctly in the first place, what is the proper way to recover from this situation with the franz-go API?

**Comment by twmb** [comment]
If I'm reading this correctly, the problem you're running into is that the transactional timeout is being hit, which is causing Kafka to bump the epoch server side, and thus when the transaction finally ends you're receiving this error.

The problem with trying to recover in any way as a client pre-2.7.0 (actually looking to be 2.8.0) is that it's impossible for the client to know if the error is the recoverable case of a transactional timeout, or the fatal case of another client starting up, using the same transactional ID, and then fencing the other. If the client tried to recover after being fenced, then they'd basically fence-loop each other and no progress would be made.

I think the safer thing to do in this case would be to bump the transactional timeout significantly to ensure that the timeout is never hit. Is this feasible?

(also if you have a lot of questions I am pretty available in the Discord)

**Comment by twmb** [comment]
Also as a side question -- the `GroupTransactSession` aborts early and often in the case of rebalances, it may be worth it to try that out (or take a look at what it does to see some risks of groups and transactions combined).

**Comment by twmb** [comment]
If neither of those work, I could add an option to allow resets on InvalidProducerEpoch, with the responsibility punted onto the user to ensure that a transactional ID will never be shared or used in other applications.

**Comment by twmb** [comment]
Also as a final note, I plan to audit the group / transactional code sooner than later just to ensure everything is solid within it :)

**Comment by dcrodman** [comment]
Appreciate the quick response! I've tried increasing the transaction timeout to values ranging from sensible to ludicrous but sadly that alone doesn't seem to have resolved it. It seems like there's some wonky interaction going on between our Sarama consumer (which already doesn't handle rebalances particularly well - https://github.com/Shopify/sarama/blob/470106f2e38132b654fbf5f802325e8b4b4a3044/consumer_group.go#L36) and this producer. Any rebalance event causes either these errors: `error beginning transaction: invalid attempt to begin a transaction while already in a transaction` or the aforementioned `INVALID_PRODUCER_EPOCH`, though I think the former is probably a fault with my implementation.

I need to rework these and will probably just start using a `GroupTransactSession` consumer in addition to these changes. I'll happily take you up on that Discord route 🙂

**Comment by twmb** [comment]
If you could try [`kcl`](https://github.com/twmb/kcl), I'm curious if `kcl consume __transaction_state` would give any clues.

I went through and checked some Kafka source at 2.3.0,
- the two places an epoch could be fenced is (1) due to a transactional timeout, (2) a second init producer ID
- the one other place an epoch is bumped is when initializing a producer ID for the first time or initializing a producer ID for a transactional ID that had an abort / commit completed

For franz-go, it only issues InitProducerID once at the start, and then any time it enters a recover state. There are logs whenever an initialization happens, if you could add WithLogger to ensure the client isn't initializing things twice.

The only way the `error beginning transaction` error is returned is if `BeginTransaction` is called before `EndTransaction` finishes. I'm struggling to find where `INVALID_PRODUCER_EPOCH` would be returned without a second client using the same transactional ID somewhere (this isn't happening, is it?)

If you can you kcl, that may shed some light on this situation as well, e.g. here's some output from my own integration tests using `kcl consume __transaction_state -g 37c22048742975827dbba8107b67a2e60c22e110fb7f6c3de2d85132cbb9f414`:
```
__transaction_state partition 32 offset 672 at [2021-02-22 16:08:16.578]
TxnMetadataKey(0) 37c22048742975827dbba8107b67a2e60c22e110fb7f6c3de2d85132cbb9f414
TxnMetadataValue(0)
      ProducerID           85
      ProducerEpoch        0
      TimeoutMillis        60000
      State                Empty
      Topics               
      LastUpdateTimestamp  1614035296578
      StartTimestamp       -1


__transaction_state partition 32 offset 673 at [2021-02-22 16:08:16.583]
TxnMetadataKey(0) 37c22048742975827dbba8107b67a2e60c22e110fb7f6c3de2d85132cbb9f414
TxnMetadataValue(0)
      ProducerID           85
      ProducerEpoch        0
      TimeoutMillis        60000
      State                Ongoing
      Topics               46c7f9efa5542d535fd2f006f6a0904805a5fe43919ca6576ccd18fa5a3297bd=>[13] 
      LastUpdateTimestamp  1614035296583
      StartTimestamp       1614035296583


__transaction_state partition 32 offset 674 at [2021-02-22 16:08:16.611]
TxnMetadataKey(0) 37c22048742975827dbba8107b67a2e60c22e110fb7f6c3de2d85132cbb9f414
TxnMetadataValue(0)
      ProducerID           85
      ProducerEpoch        0
      TimeoutMillis        60000
      State                Ongoing
      Topics               46c7f9efa5542d535fd2f006f6a0904805a5fe43919ca6576ccd18fa5a3297bd=>[13] __consumer_offsets=>[39] 
      LastUpdateTimestamp  1614035296611
      StartTimestamp       1614035296583

... and so on, as time passed
```

The specific transactional ID can be filtered for using `-g`, otherwise everything can be viewed without that.

**Comment by dcrodman** [comment]
I'm using UUIDs for the producer ID so there shouldn't be any chance of their being shared by multiple producers. This is also what we do with our current Sarama producers and none of them exhibit this problem. Is it possible that the addition of a new producer process (with a separate producer ID) on the same topic is causing the broker to expire the existing producer's epoch?

Let me take a look at `__transaction_state` with kcl and see what I can find.

**Comment by twmb** [comment]
How are you doing transactions with Sarama -- is it through manually issuing calls to brokers? I do see they have support for the transactional requests, but not much past that.

The separate producer with its distinct transactional ID producing to the topic does not cause an epoch expiration.

**Comment by dcrodman** [comment]
Sorry that was a misnomer on my part, our Sarama producers in an identical topology are not colliding on producer IDs, but since they aren't using transactions that isn't really related to the producer epoch dimension of this issue.

I'll follow up on what I find with the `__transaction_state` digging as soon as I can. Separately I'm attempting to get to the bottom of why Kafka is reporting `UNSUPPORTED_VERSION` from a different set of kgo producers despite my having set what should be the correct version. It's also still processing messages normally.

**Comment by twmb** [comment]
Do you know which API is returning UnsupportedVersion? Also are you using either of kgo.MinVersions or kgo.MaxVersions?

**Comment by dcrodman** [comment]
Yep, I'm setting both `kgo.MinVersions` and `kgo.MaxVersions` to `kversion.V2_2_0()`. Our cluster is on version 2.2.2.

I've just enabled the kgo logger in trying to figure out which API request it doesn't like. What's been interesting is there isn't a strictly 1:1 correlation between our transactions and the times Kafka reports `UNSUPPORTED_VERSION` (that is Kafka reports this error less frequently than we produce). I'm also noticing a lot of these in the logs:
```
[DEBUG] connection opened to broker; addr: <cluster>:9092, id: 0
[DEBUG] issuing api versions request; version: 3
[DEBUG] kafka does not know our ApiVersions version, downgrading to version 0 and retrying
[DEBUG] issuing api versions request; version: 0
```

**Comment by twmb** [comment]
Those three logs are fine, but I can fix the client to take into account if MaxVersions is used. With ApiVersions, it's valid to issue the highest request the client knows, and if the broker does not know how to reply, you re-issue the request with version 0. However, this would be where an `UNSUPPORTED_VERSION` comes from. Is the error being bubbled to a client somewhere or are you seeing it in Kafka logs?

**Comment by twmb** [comment]
I've pushed https://github.com/twmb/franz-go/commit/89940858a373d38388893614e47ab149d1d41933 to ensure ApiVersions is only issued with the user specified max version, if user specified versions exist.

**Comment by dcrodman** [comment]
I can't find any occurrences of it bubbling up to the client, we're seeing this from the Prometheus metrics we export via https://raw.githubusercontent.com/prometheus/jmx_exporter/master/example_configs/kafka-2_0_0.yml. It shows up in `kafka_network_requestmetrics_errors_total`

**Comment by dcrodman** [comment]
Awesome, thanks! I'll try that out right now and see if it resolves these

**Comment by dcrodman** [comment]
Confirmed that https://github.com/twmb/franz-go/commit/89940858a373d38388893614e47ab149d1d41933 solved the `UNSUPPORTED_VERSION` errors! Thanks for fixing that so quickly

**Comment by twmb** [comment]
No worries, thanks for confirming! Let me know if you have any luck with `kcl`, I'm pretty interested in knowing what's going on with these transactions.

**Comment by dcrodman** [comment]
There isn't a verbose mode for `kcl consume` is there? Running `kcl consume __transaction_state -X seed_brokers=...`  is just leaving it hanging, I'm running this within a Kubernetes cluster and am assuming it's timing out somewhere

**Comment by twmb** [comment]
You can use `--log-level=debug` to print out the debug logs from the kgo library itself, although those are a bit sparse and may not tell everything.

**Comment by dcrodman** [comment]
Actually that's exactly what I needed to sort this out - it was just spinning on connection errors:
```
[WARN] unable to open connection to broker; ...
```

I believe I've figured out at least part of the problem. In keeping with our usual patterns I'm initializing a `sarama.ConsumerGroupHandler` with an instance of `kgo.Producer` for the Kafka operations. This in turn gets passed to [ConsumerGroup.Consume](https://github.com/Shopify/sarama/blob/470106f2e38132b654fbf5f802325e8b4b4a3044/consumer_group.go#L44), which spins off a [separate goroutine per claim](https://github.com/Shopify/sarama/blob/470106f2e38132b654fbf5f802325e8b4b4a3044/consumer_group.go#L27). I'm not 100% certain of the exact interaction but from the kgo logs it's become obvious that sharing this producer instance across goroutines is clobbering the transaction state, especially on cancellation.

Whether or not it _fully_ resolves my problems it looks like I need to rework this a little bit - either allowing each `Handle(...)` call to create its own producer (if that's not prohibitively expensive or too much load on the brokers) or just writing messages from all the claim consumers back to a single producer thread that processes each batch in its own transaction.

**Comment by twmb** [comment]
Interesting, glad you were able to narrow it down. The producer definitely can't be shared across all "claim" goroutines. The beginning, flushing, committing offsets, and ending of transactions needs to be owned by one goroutine. I think that an easy model would be to send all received messages from sarama down a channel to a dedicate goroutine that does producing may be a viable model. Sarama's model of having a goroutine per assigned partition doesn't really mesh well with what Kafka does.

Out of curiosity, is there a reason to now use full EOS? Using a consumer group and then producing with transactions I think gets most of the way there, but the group itself can be transactional as well.

**Comment by dcrodman** [comment]
I'll probably do something like that, the "batches" are significant and I have some ideas for delineating those to indicate the begin/end transaction calls. 

This is the only node in our pipeline that isn't strictly speaking guaranteed to be exactly-once, so I've been iterating on the "riskiest" parts. I'll be following up the producer transactions with consumers built on `GroupTransactSession` to reach my goal of making the entire thing transactional as you said

**Comment by dcrodman** [comment]
As I'm exploring this full EOS route, what's the proper way to accomplish this with kgo? Should I be using one client for both the `GroupTransactSession` and the producer given the input and output topics are on the same cluster? If so, should I be using `client.EndTransaction` or `session.End` (assuming the former)?

**Comment by twmb** [comment]
The proper way would be to use the session for beginning and ending, and then the client that you used for the session for producing. I could add the Produce function to the GroupTransactSession to make that a bit more obvious.

One example can be seen in kcl:
The client polls, reads fetches, execs the ETL program:
https://github.com/twmb/kcl/blob/2a7df6876dcc07954d98be7a1dd2dc3a5fae1550/commands/transact/transact.go#L214-L279
Then we begin a session, produce, and end the session in a tight smaller block:
https://github.com/twmb/kcl/blob/2a7df6876dcc07954d98be7a1dd2dc3a5fae1550/commands/transact/transact.go#L281-L307

The main benefit of the transact session is to abort early and often in the face of rebalances and whatnot, which helps reduce the risk of what KIP-588 is intended to address.

**Comment by dcrodman** [comment]
Okay excellent thank you, let me see if I can get this working. The behavior of `GroupTransactSession` with respect to rebalances is specifically what I'm after, this gets tricky with the timeouts and our consumers can take a long time to process a message. So just aborting after the claims are revoked is loads safer.

A `Produce()` function on the session would be helpful, it admittedly took me a little while to piece together exactly how this needs to work and that's with having a decent sense of where to start. Just as a suggestion, tossing in some docs or an example for this (probably common) use of an ETL EOS consumer would probably make this a bit more approachable.

**Comment by twmb** [comment]
(in the prior comment, I meant KIP-447, not KIP-588)

The approach the GroupTransactSession takes is to immediately set the internal state to "revoked" once a revoke happens, and then call the user revoke. When the session ends, `End` commits offsets, and if the state is still not revoked, attempts to end the transaction successfully, otherwise aborts. 

I spent some time looking into Kafka source to ensure I'm handling the commit properly in the face of a group rebalance ending _after_ the TxnOffsetCommit but _before_ the EndTxn. I'm not entirely sure I am. I'm going to run a local test shortly to ensure this is being handled properly.

I'll probably add `Produce` to the transact session. I've deliberately gone with a more minimal API to start to help ensure I break less code in the future if an API needs to change, but I'm pretty happy with the GroupTransactSession. I'm less happy with any consume group offset API.

**Comment by dcrodman** [comment]
(Sorry I was out at the end of last week and am just getting back on this.) I've rewritten the ETL worker so that it's using a `GroupTransactSession` and a transactional producer, fanning out a goroutine per partition. Each goroutine writes any messages back to a channel while the producer reads messages from that channel, `Send`s everything in one transaction, and `txnSess.End`ing after all of the handlers have burned through their messages. In essence, parallelizing the processing of any assigned claims but all sharing one producer to avoid the issue I'd encountered previously with sharing a producer across the handlers.

I was still encountering the `INVALID_PRODUCER_EPOCH` errors but the kgo logs helped me figure out that it was because the consumer was returning 6000+ records in a batch and because each record can take a few seconds the transaction timeout was being exceeded. I think for my purposes I just need to bump that timeout and set a smaller `FetchMaxBytes` so that the consumer will retrieve fewer records at once.

**Comment by dcrodman** [comment]
Exactly, it made more sense in my case to just replace the entire Sarama consumer group. Great thanks for the tip, I'm working out the proper value for `FetchMaxBytes`. I haven't encountered that issue but I'll keep an eye on it, this hasn't been anywhere near production yet but at least if I run into I know where to look

**Comment by dcrodman** [comment]
Going to go ahead and close this since we got to the root of my original problem and I've moved over to the Discord. Thank you!

**Comment by twmb** [comment]
For posterity, this was moved to discord and addressed with the following commit and released in v0.6.10:
- https://github.com/twmb/franz-go/commit/2dc7d3f
Basically, a transaction was started because some stuff was consumed even though nothing was produced. The client did not track that a transaction was started in this case and avoided issuing a commit. Now, we do so.
